fern/src/vm.rs

```diff
@@ -4,30 +4,44 @@ use std::vec;
 
 use crate::{
     arch::{
-        InstructionAddress, LocalAddress, Word, ARGUMENT_LOCALS, CODE_SIZE,
-        LOCALS_SIZE, RETURN_LOCALS,
+        InstructionAddress, LocalAddress, Word, ARGUMENT_LOCALS, LOCALS_COUNT,
+        MAX_CODE_LENGTH, RETURN_LOCALS,
     },
-    opcode::{Op, IMM_BITS, IMM_EXT_BITS},
+    op::{Op, IMM_BITS, IMM_EXT_BITS},
 };
 
-macro_rules! sign_extend_to {
-    ($t:ty, $value:expr, $bits:expr) => {{
-        let value = $value as $t;
-        let sign_bit = 1 << ($bits - 1);
-        if value & sign_bit != 0 {
-            let extension = !((1 << $bits) - 1);
-            value | extension
-        } else {
-            value
-        }
-    }};
+fn sign_extend_to<
+    In: num_traits::Unsigned + num_traits::PrimInt + num_traits::AsPrimitive<Out>,
+    Out: 'static + num_traits::Unsigned + Copy,
+>(
+    value: In,
+    bits: usize,
+) -> Out {
+    let sign_bit = In::one() << (bits - 1);
+    if value & sign_bit != In::zero() {
+        let extension = !((In::one() << bits) - In::one());
+        value | extension
+    } else {
+        value
+    }
+    .as_()
 }
 
 struct StackFrame {
-    locals: [Word; LOCALS_SIZE],
+    // TODO: do we wnat this on the heap?
+    locals: [Word; LOCALS_COUNT],
```

Comment:
I think it's small enough that the stack is fine but we can think about it in the future